traffic_ops/traffic_ops_golang/deliveryservice/servers/servers.go

```diff
@@ -263,7 +243,409 @@ func (dss *TODeliveryServiceServer) Delete(db *sqlx.DB, user auth.CurrentUser) (
+			return
+		}
+
+		// perform the insert transaction
+		rollbackTransaction := true
+		tx, err := db.Beginx()
+		defer func() {
+			if tx == nil || !rollbackTransaction {
+				return
+			}
+			err := tx.Rollback()
+			if err != nil {
+				log.Errorln(errors.New("rolling back transaction: " + err.Error()))
+			}
+		}()
+
+		if err != nil {
+			log.Error.Printf("could not begin transaction: %v", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		// We have to get the server Ids and iterate through them because of a bug in the Go
+		// transaction which returns an error if you perform a Select after an Insert in
+		// the same transaction
+		for serverIds.Next() {
+			var serverId int
+			err := serverIds.Scan(&serverId)
+			dtos := map[string]interface{}{"id":dsId, "server":serverId}
+			resultRows, err := tx.NamedQuery(insertIdsQuery(), dtos)
+			if err != nil {
+				if pqErr, ok := err.(*pq.Error); ok {
+					err, eType := dbhelpers.ParsePQUniqueConstraintError(pqErr)
+					log.Error.Printf("could not begin transaction: %v", err)
+					if eType == tc.DataConflictError {
+						handleErrs(http.StatusInternalServerError, err)
+						return
+					}
+					handleErrs(http.StatusInternalServerError, err)
+					return
+				}
+				log.Errorf("received non pq error: %++v from create execution", err)
+				return
+			}
+			resultRows.Next()
+		}
+
+		err = tx.Commit()
+		if err != nil {
+			log.Errorln("Could not commit transaction: ", err)
+			return
+		}
+		rollbackTransaction = false
+
+		// marshal the results to the response stream
+		tcPayload := tc.DeliveryServiceServers{payload.ServerNames, payload.XmlId}
+		payloadResp := tc.DSServersResponse{tcPayload}
+		respBts, err := json.Marshal(payloadResp)
+		if err != nil {
+			log.Errorln("Could not marshal the response as expected: ", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		w.Header().Set(tc.ContentType, tc.ApplicationJson)
+		fmt.Fprintf(w, "%s", respBts)
+		return
+	}
+}
+
+func selectDeliveryService() string {
+	query := `SELECT id FROM deliveryservice WHERE xml_id = $1`
+	return query
+}
+
+func insertIdsQuery() string {
+	query := `INSERT INTO deliveryservice_server (deliveryservice, server) 
+VALUES (:id, :server )`
+	return query
+}
+
+func selectServerIds() string {
+	query := `SELECT id FROM server WHERE host_name in (?)`
+	return query
+}
+
+// api/1.1/deliveryservices/{id}/servers|unassigned_servers|eligible
+func GetReadHandler(db *sqlx.DB, filter string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		handleErrs := tc.GetHandleErrorsFunc(w, r)
+		params, err := api.GetCombinedParams(r)
+		if err != nil {
+			log.Errorf("unable to get parameters from request: %s", err)
+			handleErrs(http.StatusInternalServerError, err)
+		}
+
+		where := `WHERE s.id in (select server from deliveryservice_server where deliveryservice = $1)`
+
+		if filter[0] == 'u' {
```

Comment:
This will panic if filter is empty. Should be `if len(filter) > 0  && filter[0] == 'u' {`

Reply:
Ah, routes is passing a string constant. Using magic strings is error-prone, if someone mistypes the string. A boolean or an enum would catch that at compile time. 

I'd suggest a boolean, because it's easier. If you want "pass by name" semantics, you can define consts for it, `const Assigned = true; const Unassigned = false`, and in `routes.go`, `dsserver.GetReadHandler(d.DB, dsserver.Unassigned)`.